Packets working branch

CollinValley opened the pull request:
This is a huge commit, everyone please go through it and point out anything that is hacky, could use improvement. Particularly, I want to weed out things that likely don't do what you expect. I want this library to not easily panic, and have a what you see is what you get feel to it.  This is a big ol' boy, and I'm going to want to merge it not completely perfect, but I think it is in a functional enough state now that it could be used to begin building elements.  Once we have a list of fixes etc, I'll pare down and clean up the commit messages.

CollinValley committed:
Packet: A route-rs packet library

This is the base commit of the route-rs packet library. It is at a
point where it is not ready to be merged to master, but is ready to
being accepting PRs and outside help. Right now, there exists a unique
structure for each relevent packet type, checked constructors for all
of them, and conversion functions everywhere it makes sense. The code
is not well tested, and does not yet conform to large sections of the
RFC requirements. This is why it is living in seperate branch for now.

Until it is ready to submit to master for an intensive CR, I am going
to install myself as the BDFN (Benevelent Dictator for now) and accept
PRs solving the various issues detailed in the sprint. I will CR all
the changes myself. I am trying to move fast on this library since there
is a lot of ground to cover to get it to a functional stage, as such I
will only sometimes request CR/outside review for various changes I
make. Others will have a change to voice concerns and advocate for
changes either through submitting PRs, or during the CR process for
submitting to master.

CollinValley committed:
Ipv4: Add SRC and DEST Address helpers

Adds src_addr, dest_addr, set_src_addr, and set_dest_addr helper
functions to the Ipv4Packet struct. Also creates a new common Ipv4Addr
structure that gets returned to the user when they call the getter.

CollinValley committed:
Ethernet: Improve code for MacAddr access

This uses less ugly array semantics for MacAddr, still has magic
numbers in it though.

CollinValley committed:
Ipv4: Add header_len function, properly calculate payload offset

CollinValley committed:
Ipv4: Implement function to return Options field

This provides an Option<Cow<[u8]>> of the options field. Since it is
not garunteed to exist if there are no options. Plus it is a great pun.

Also I ran cargo fmt for the first time in awhile.

CollinValley committed:
Declare IpProtocol Enum, uggh it is long. Also allow dead code where it makes sense

CollinValley committed:
Add IpProtocol From<u8> impl, wow this sucks, but I never have to do it again :)

CollinValley committed:
Ipv4: Add protocol and total_length functions

CollinValley committed:
Fix small clippy warnings, like module inception, and some for loop removal

CollinValley committed:
Replace clone_from_slice with copy_from_slice since it is faster with u8's

CollinValley committed:
Ipv4, add handful of basic functions. Also add checksum capabilities

CollinValley committed:
EthernetFrame: Create basic test suite

Lays out a series of basic tests for the methods available for
EthernetFrames. Tests setting all the field, creating from new,
accessing payloads, and confirming that errors are returned when an
invalid data buffer is provided to the new function.

CollinValley committed:
Ipv4Packet: Implement basic test

Adds basic test to create an Ipv4 packet from an EthernetPacket.
Checks that all the helper functions to access the values return a
sane result. Also improves quality of life by defining a Result type
for Ipv4Packet conversions.

CollinValley committed:
Packets: Add Result type to improve QoL

This simplifies syntax by having a result type for all conversions.
Which prevents needing the turbofish every time we wish to convert a
packet from one type to another.

CollinValley committed:
Ipv4Packet: Improve basic test syntax

CollinValley committed:
Ipv4: Fix checksum math and add tests

While writing the tests, woo, I realized I had previously been doing my
addition math incorrectly when creating the checksum. This lead to
incorrect results. I have since corrected that issue, and now the
checksum related functions work as intended.

CollinValley committed:
IPv4Packet: Improve syntax of checksum code

CollinValley committed:
Ipv4Packet: Small function cleanup.

CollinValley committed:
Ipv6Addr: Create struct and helper functions

CollinValley committed:
Ipv6: Create more helper functions.  Cleanup comments

CollinValley committed:
IPv4, small syntax suger change

CollinValley committed:
Ipv6: Add more helper functions.

Add payload, src_addr, dest_addr, and a few other helper functions
to Ipv6. Along the way also adds a helper function to Ipv6Addr to
make setting new Ipv6 addresses easier in the libary.

CollinValley committed:
Ipv6: Add tests for basic packet commands

Adds test for basic Ipv6 packet, verifies all fields are correct.
Also add test for setting the source and destination addresses with
the new Ipv6addr structure.

CollinValley committed:
IPLayer: Add extension headers support to Ipv6

Also fixup set_payload in Ipv4 to be a bit more idiomatic.

CollinValley committed:
Ipv6: Test Set Payload

Adds test and resolves a small error in the set payload function

CollinValley committed:
Ipv6: Add set_extension headers to Ipv6

CollinValley committed:
Changes from cargo fmt

CollinValley committed:
Ipv4: Implement set_options

Add set options function, allowing users to manually set the options
field. Also made some changes to the internal repreasentation of the
Ipv4 packet, so that we keep track of the index payload offset, rather
than the confusing header length.

CollinValley committed:
Introduce UDP functions, change some small behavior with Ipv4

CollinValley committed:
Ipv4Packet: Add packet offset

Adds packet offset to Ipv4Packet to remove magic, and in the future
allow frame headers to be a length other than 14.

CollinValley committed:
Ipv6Packet: Add packet offset

Adds support for variable frame length(other than 14), and removes
a good deal of magic numbers from the source base.

CollinValley committed:
Packets: Adjust structs to carry offset data

Adds values to each struct so that we retain offsets for the start of
each layer's header. This is useful when creating a less specific
headeer from a more specific one. For instance, when we move UDP->Ipv4.
It is easy in the from function to reference segments packet_offset
field. This means that creating a segment from scratch requires
providing other values, like the packet offset. But this helps the code
remain agnostic to the various layer's thicknesses.

CollinValley committed:
Packets: Add helper function to determine segment

Add two helper functions for Ipv4 and Ipv6 to determine the IpProtocol
type of their payload. These functions are useful since they can be
called when creating a new segment, TCP or UDP, so that the new function
may check that the provided data does in fact say it contains the
appropriate transport type.

CollinValley committed:
Ipv4: Small fixup, resolve payload() bug

CollinValley committed:
tcp: Implment TCP

Implements the basics for TCP, and also adds  a test module
asserting that all the basics work.

CollinValley committed:
UDP: Add tests for UDP

Adds basic test and resovles a small bug found by the test.

CollinValley committed:
Run clippy & fmt, fix any issues

moosingin3space commented on route-rs-runtime/src/lib.rs:
```diff
@@ -23,6 +23,8 @@ pub mod link;
 /// program takes the user provided graph, and generates a pipeline. It connects input and output channels to the pipe, so that it can be
 /// hooked up to packet sources and sinks, and drops the pipeline into the runtime. Generally can be abstracted to the concept, router.
 pub mod pipeline;
+pub mod packet;
```
I think this could work better as a standalone crate -- that way I could use it in `afpacket` tests, for instance.

CollinValley replied:
I want to agree, would be nice to version the library too, so I can tell users it is 0.1 lol.  Will create a patch to move the files.

CollinValley commented on route-rs-runtime/src/packet/ipv6.rs:
```diff
@@ -0,0 +1,336 @@
+}
+
+impl<'packet> Ipv6Packet<'packet> {
+    fn new(packet: PacketData) -> Result<Ipv6Packet, &'static str> {
```
Forgot to switch this so the new function takes the variable packet_offset length. Will get that tomorrow.

CollinValley replied:
See ipv4::new() for the idea.

moosingin3space commented on route-rs-runtime/src/packet/ethernet.rs:
```diff
@@ -0,0 +1,113 @@
+}
+
+impl<'frame> EthernetFrame<'frame> {
+    pub fn new(frame: PacketData) -> Result<EthernetFrame, &'static str> {
```
Let's think about a legit `Error` type. I'm not too sure what the best way to do `Error`s is since `failure` became superseded by the stdlib though.

CollinValley replied:
Yeah, I would like to be able to return formatted strings or something.

CollinValley replied:
I'll take a look into Error types. Ideally we could return formatted strings

moosingin3space commented on route-rs-runtime/src/packet/ethernet.rs:
```diff
@@ -0,0 +1,113 @@
+
+    // This is super uggo but I don't care right now, I'm sure this is a cleaner way to do it.
+    pub fn set_dest_mac(&mut self, mac: MacAddr) {
+        self.data[0] = mac.b0;
```
If `MacAddr` stored an array internally, you could use [this](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice).

CollinValley replied:
Yeah, I will switch it over, similarly to how I did Ipv4/Ipv6. They keep an array internally.

moosingin3space commented on route-rs-runtime/src/packet/ethernet.rs:
```diff
@@ -0,0 +1,113 @@
+    }
+}
+
+impl<'packet> From<TcpSegment<'packet>> for Result<EthernetFrame<'packet>, &'static str> {
```
Any reason for `From` instead of `TryFrom` here?

CollinValley replied:
/facepalm this is exactly what I wanted when I made the Result.  https://doc.rust-lang.org/std/convert/trait.TryFrom.html Will convert that over tonight.

moosingin3space commented on route-rs-runtime/src/packet/ipv4.rs:
```diff
@@ -0,0 +1,72 @@
+    fn new(packet: PacketData) -> Result<Ipv4Packet, &'static str> {
+        //Header of Ethernet Frame: 14 bytes
+        //Header of IPv4 Frame: 20 bytes
+        if packet.len() < 14 + 20 {
```
Ideally a packet type doesn't have to worry about what its containing packet type is. This is worth discussion.

CollinValley replied:
This is an old commit. In the latest I added a packet_offset parameter that means the packet can be agnostic to the amount of data preceeding it in the array.

moosingin3space commented on route-rs-runtime/src/packet/packet.rs:
```diff
@@ -0,0 +1,23 @@
+//Let's use this area for now to declare common structs, constants, and common helper functions.
+#[derive(Eq, Clone, Copy, Hash, PartialEq)]
+pub struct MacAddr {
```
See above comments about `copy_from_slice`.

moosingin3space commented on route-rs-runtime/src/packet/udp.rs:
```diff
@@ -0,0 +1,70 @@
+use crate::packet::*;
+
+pub struct UdpSegment<'packet> {
```
Not sure "segment" is the semantically-correct name for this -- `UdpDatagram` maybe?

CollinValley replied:
In TCP they are called Segments, UDP, doesn't seem like it has a defined term. Some people say segment, others say packet.

moosingin3space reviewed (changes requested):
I like this API. That said, it'd be much easier to read with constants, and I'd like to have a discussion about ways we can potentially make each inner packet independent of the outer packet storing it.

CollinValley commented:
> I like this API. That said, it'd be much easier to read with constants, and I'd like to have a discussion about ways we can potentially make each inner packet independent of the outer packet storing it.

Yeah,  I made some strides on this front by retaining offsets as we go up stack.  A TCPsegment keeps track of the IP Header offset, it's own TCP Header offset, an then it's payload offset. This allows us to remain more agnostic about why layer is beneath us. Unfortunately, there are instances of coupling, like the combined checksum in TCP, that break the model.  Also the fact that TCP has no length field, it just uses the IP length. Super annoying.

moosingin3space commented on route-rs-runtime/src/packet/types.rs:
```diff
@@ -0,0 +1,427 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            " {:X}::{:X}::{:X}::{:X}::{:X}::{:X} ",
```
Switch to single colon.

CollinValley replied:
Done

moosingin3space commented on route-rs-runtime/src/packet/types.rs:
```diff
@@ -0,0 +1,427 @@
+
+//Most significant byte is 0th
+#[derive(Eq, Clone, Copy, Hash, PartialEq, Debug)]
+pub struct Ipv4Addr {
```
This already exists in the stdlib.

moosingin3space commented on route-rs-runtime/src/packet/types.rs:
```diff
@@ -0,0 +1,427 @@
+}
+
+#[derive(Eq, Clone, Copy, Hash, PartialEq, Debug)]
+pub struct Ipv6Addr {
```
See above.

moosingin3space commented on route-rs-runtime/src/packet/ipv6.rs:
```diff
@@ -0,0 +1,336 @@
+                }
+                header_ext_len = data[offset + 1];
+                if header_ext_len == 0 {
+                    //fragments have the minimum of 8, but it set to zero for some dumb reason
```
Can you cite your source here please?

moosingin3space commented on route-rs-runtime/src/packet/ipv6.rs:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    //TODO: Test the get and set for extension headers.
+    pub fn extension_headers(&self) -> Vec<Cow<[u8]>> {
```
Use a common function (between this and `get_ipv6_payload_type`)

CollinValley replied:
Took at look at this, while a good amount of the logic is the same. The main loop differs which makes abstracting the logic hard.

CollinValley committed:
Packets: Change From to TryFrom where relevant

CollinValley committed:
Packets:  Use try_into instead of Array literals

In the library we were populating from_be_bytes with an inline
instantiated array literal, now we are using a try_into call from a
slice. This is more idiomatic and works better for larger arrays.

CollinValley committed:
route-rs-packets: Move packets to its own crate

Moves packets to its own crate, named route-rs-packets. Able to be
used by anyone now.

CollinValley committed:
Add license file

CollinValley committed:
Fix mac address printing to use single colon

CollinValley committed:
Packets: Switch from home-rolled IpAddr to STD

The standard library has Ipv4/Ipv6 addresses that work for our
purposes, switch to using those instead of the ones we defined.

CollinValley committed:
Packets: Cleanup comments, add source for fragement

CollinValley committed:
run cargo fmt

CollinValley committed:
Clippy fixup

CollinValley commented:
I rebased this against master, so it once I have a little bit of approval we can merge this in. I think it is in a useful state right now, enough that I can call it version 0.1 It should also unblock us from building a router/elements using it.

eeverett6 commented on route-rs-packets/src/ipv4.rs:
```diff
@@ -0,0 +1,344 @@
+}
+
+/// Returns Ipv4 payload type, reads the header information to get the type
+/// of IpProtocol payload is included. Upon error, returns IpProtocol::Reserved.
```
Why not an error type? What if this reserved value is later assigned?

CollinValley replied:
Idk making an error type is kind of annoying, this is quick and dirty. Error type is on the list of improvements, i can spawn off an issue.

CollinValley replied:
I can fix this by returning a Result at least to start.

moosingin3space replied:
What kind of errors are possible?

CollinValley replied:
There are a few numbers that are undefined?

CollinValley replied:
Edit: Nevermind this error makes sense. The only error is that they didn't hand me something that could have been an Ipv4 packet.

I am going to wrap this in a Result rather than a custom error to be more clear.

eeverett6 commented on route-rs-packets/src/ipv6.rs:
```diff
@@ -0,0 +1,352 @@
+}
+
+/// Returns Ipv6 payload type, reads the header information to get the type
+/// of IpProtocol payload is included. Upon error, returns IpProtocol::Reserved.
```
Same question about reserved here

CollinValley replied:
ack

eeverett6 commented on route-rs-packets/src/types.rs:
```diff
@@ -0,0 +1,336 @@
+    Reserved,
+}
+
+impl From<u8> for IpProtocol {
```
Couple thoughts:
- It feels wordy to have to do this twice
- How do we turn an IpProtocol into a u8 when writing out a packet/overwriting the protocol field?

Have you seen the solution in libpnet? https://github.com/libpnet/libpnet/blob/master/pnet_packet/src/ip.rs It looks like the protocol is a struct with a single u8 field, then there are two mappings, one a list of constants, and one a display function.

CollinValley replied:
Hmm this is a decent solution.  I'd be fine with changing it to something like libpnet's

moosingin3space replied:
Yeah, arguably `From<T>` isn't what we want here -- let's make clients explicitly convert integers to the typed values.

CollinValley replied:
Going to shelve this for after merge cleanup.

CollinValley replied:
#178

eeverett6 commented on route-rs-packets/src/ethernet.rs:
```diff
@@ -0,0 +1,169 @@
+    pub fn new(frame: PacketData) -> Result<EthernetFrame, &'static str> {
+        // Ethernet II frames must be at least the header, which is 14bytes
+        // 0                    6                    12                      14
+        // |---6 byte Dest_MAC--|---6 byte Src_MAC---|--2 Byte Payload Len---|
```
In Ethernet II, the 12th and 13th bytes are actually the ethertype field, not a payload length. I believe there is no length field in ethernet II. We will probably need something like the IpProtocol types for EtherTypes as well

CollinValley replied:
Interesting.  https://en.wikipedia.org/wiki/Ethernet_frame Looks like it is EtherType in the case it is an Ethernet II frame. As far as I'm aware, we really only expect Ethernet II types on the internet?  https://en.wikipedia.org/wiki/EtherType Looks like it is fairly complex.

moosingin3space replied:
EtherTypes are how we define what the next packet type is, though -- since LLDP and ARP don't use IP headers. I'm not sure we can avoid it. According to https://en.wikipedia.org/wiki/Ethernet_frame#Types, it looks like Ethernet II is the most common, with raw IEEE 802.3 being an ancient Novell thing.

eeverett6 replied:
I would honestly just drop support for the 802.3 stuff entirely and just support Ethernet II with the EtherType field. I don't think it's worth it to be generic across so many different standards when we'll never encounter them.

CollinValley replied:
We chatted offline about this, my next push will shift us to just supporting EtherType instead of length field. I will still check that it is in a valid EtherType range.

eeverett6 commented on route-rs-packets/src/ipv4.rs:
```diff
@@ -0,0 +1,344 @@
+        )
+    }
+
+    pub fn dcsp(&self) -> u8 {
```
dscp

CollinValley replied:
ack

eeverett6 commented on route-rs-packets/src/ethernet.rs:
```diff
@@ -0,0 +1,169 @@
+
+impl<'packet> TryFrom<Ipv4Packet<'packet>> for EthernetFrame<'packet> {
+    type Error = &'static str;
+
```
Should we have a way of encapsulating an Ipv4 packet if it didn't originally come from an ethernet frame?

CollinValley replied:
In this case, this function creates an EthernetFrame from an Ipv4 or Ipv6.  So it isn't really encapsulating per se, just a type conversion.  We could add support for other Layer 2 protocols if need be, since the try_from would fail in that caseif you tried to make an ethernet frame.

eeverett6 replied:
I'm thinking of the case where say a processor/element creates an Arp Message, that now needs to be put behind an Ethernet header to be sent out over the wire. These would be separate from the TryFroms, but could be methods that take an arbitrary payload, a src/dest MAC addr, and an ethertype, then replaces the old l2 header or pushes more data onto the front of the packet as necessary

eeverett6 reviewed (commented):
Didn't have time to look over the whole thing tonight, but posting my feedback so far. I should be able to contribute commits to address some of this feedback myself soon too.

CollinValley commented:
Thanks for looking through it Ethan, obviously any contribution is welcome.  I'll start trying to address the simple things, and will create some issues for the stuff that will take longer.

CollinValley committed:
Create result type for get ip payload functions

CollinValley committed:
Packets: Ethernet: Fix EtherType bug

Previously, we were not handling the case where the field proceeding the
two MAC addresses, generally called the EtherType field, is valid as
either the payload_len or an EtherType value. Which it is depends on the
value present. Generally, in modern times, it is used to indicate that
the frame contains an IPv4 packet.

We now handle this case as simply as possible, with a check that the
field is inside a valid range, and checking for valid length fields
where it makes sense.

We also change the set_payload function where it only updates the
ether_type field if it currently is being used as a payload_len. In the
future, I suspect this area will need to be built out more to support
VLAN tagging, but that is pretty far down the road.

CollinValley commented:
Just resolved the last issue from the previous CRs. Also rebased the branch against master to ensure everything is still working as expected.

eeverett6 commented on route-rs-packets/src/types.rs:
```diff
@@ -0,0 +1,348 @@
+    WESP,
+    ROHC,
+    Unassigned,
+    Use_for_expiramentation_and_testing,
```
experimentation

CollinValley replied:
whoops, fixed.

CollinValley committed:
Only support EtherType, don't validate the field yet though, since we don't really have anything to do with it yet

ddowl reviewed (approved):
lol I read this
